src/app/autocomplete-input/autocomplete-input.component.ts

```diff
@@ -0,0 +1,62 @@
+import {Component, Input, Output, ChangeDetectionStrategy, EventEmitter} from '@angular/core';
+import { SchemaKeysStoreService } from '../shared/services/schema-keys-store.service';
+import {Observable} from 'rxjs/Observable';
+
+@Component({
+  selector: 'me-autocomplete-input',
+  templateUrl: './autocomplete-input.component.html',
+  styleUrls: ['./autocomplete-input.component.scss'],
+  changeDetection: ChangeDetectionStrategy.OnPush
+})
+export class AutocompleteInputComponent {
+
+  @Output() valueChange: EventEmitter<string> = new EventEmitter<string>();
+  @Input() value = '';
+  @Input() styleClass = '';
+
+  splitToken = '/';
+  currentPath = '';
+  dataSource: Observable<any>;
+
+  private _typeaheadNoResults = false;
+
+  constructor(private schemaKeysStoreService: SchemaKeysStoreService) {
+    this.dataSource = Observable
+    .create((observer: any) => {
+      observer.next(this.value);
+    })
+    .mergeMap((token: string) => this.getStatesAsObservable(token));
+  }
+
+  getStatesAsObservable(token: string): Observable<any> {
```

Comment:
doing all this logic in this method is too much. I suggest you do a `map` before `mergeMap` that takes `token` that returns `path` and `lastSeparatorIndex` then a [do](https://www.learnrxjs.io/operators/utility/do.html)  that assigns the` path` to` currentPath` then another `map` finally returns the `queryRegExp`

After all of this, this function would be only `return` statement.